### PR TITLE
Support editing data files in sub-directories

### LINKED ISSUE #302: Ability to detect datafiles in sub-directories
Now that `documents` and `pages` within sub-directories can be viewed and edited, its odd that data files within sub-directories are still hidden away.

Possible use-case:
```
./_data
├── locales
| └── en.yml
| └── fr.yml
```

### PR DESCRIPTION
**Pull request by ashmaroli**
Fixes #302 
- [x] Update API to support data files in sub-dirs
- [x] Update Front End
- [x] Update documentation

### API Breaking Changes
  - `path`, `relative_path`, `absolute_path` is now derived from data file id
  - `'path' != 'relative_path'`
  - `relative_path` is now relative to `data_dir` and does not have a leading `/`
  - file `id` now has a `.yml` extension even if the assoc. file doesn't.

### CODE REVIEW

**Comment by ashmaroli** [comment]
### Front-end Changes
- CRUD data files recursively
- Added ability to rename data files. Rename their `path` to move a file from one directory to another.

**Comment by ashmaroli** [comment]
@benbalter This is ready for your review. 
The `relative_path` now outputs path relative to the `data_dir` and it doesn't have a leading `/` to easily join with directory paths that have a trailing `/` to differentiate itself from a path to an id without an extension.

File: src/components/Breadcrumbs.js [hunk 1]
```diff
@@ -22,10 +22,17 @@ export default class Breadcrumbs extends Component {
       const paths = splat.split('/');
       links = _.map(paths, (path, i) => {
         const before = (i == 0) ? '' : paths.slice(0, i).join('/') + '/';
-        return {
-          href: `${base}/${before}${path}`,
-          label: path
-        };
+        if (type == "data files") {
```
Comment (mertkahyaoglu) [inline src/components/Breadcrumbs.js]: Can't we include the trailing slash in `path` variable like we do with other types?

File: src/components/Breadcrumbs.js [hunk 2]
```diff
@@ -12,7 +12,7 @@ export default class Breadcrumbs extends Component {
     let base;
     if (type == 'pages') {
       base = `${ADMIN_PREFIX}/pages`;
-    } else if (type == 'datafiles') {
+    } else if (type == 'data files') {
```
Comment (mertkahyaoglu) [inline src/components/Breadcrumbs.js]: Why rename?

File: src/containers/views/DataFileEdit.js [hunk 3]
```diff
@@ -47,23 +63,36 @@ export class DataFileEdit extends Component {
     }
   }
 
+  getDirectoryfromPath(path) {
```
Comment (mertkahyaoglu) [inline src/containers/views/DataFileEdit.js]: You can retrieve the directory via;

```javascript
const [directory, ...rest] = params.splat;
```

File: src/components/Breadcrumbs.js (hunk 2 above)
Comment (ashmaroli) [inline src/components/Breadcrumbs.js]: I had included the explanation for this in the commit message:
```
[...] to render into 'Data Files' instead of 'Datafiles'
```

File: src/containers/views/DataFileEdit.js (hunk 3 above)
Comment (ashmaroli) [inline src/containers/views/DataFileEdit.js]: yes I've used that in a few places elsewhere.

File: src/components/Breadcrumbs.js (hunk 1 above)
Comment (ashmaroli) [inline src/components/Breadcrumbs.js]: are you referring to the pre-API variable or `path` within `Breadcrumbs`? If its the latter, the reason is that variable `before` uses `path` to generate its value and I get the following:
```
Data Files / movies/ / genres/ /
```

Comment (mertkahyaoglu) [inline src/components/Breadcrumbs.js]: Trailing slash should be optional for the API. For example while `/_api/data/movies/` returns a result  , `/_api/data/movies` doesn't.

File: src/components/form/InputPath.js [hunk 4]
```diff
@@ -16,12 +16,12 @@ export default class InputFilename extends Component {
     if (type == 'posts') {
       const date = moment().format('YYYY-MM-DD');
       placeholder = `${date}-your-title.md`;
-    }else if (type == 'datafiles') {
-      placeholder = 'your-filename.yml';
+    }else if (type == 'data files') {
+      placeholder = '<directory/>your-filename.yml';
```
Comment (mertkahyaoglu) [inline src/components/form/InputPath.js]: Typo here => `<directory> / your-filename.yml`

File: src/components/Breadcrumbs.js (hunk 1 above)
Comment (ashmaroli) [inline src/components/Breadcrumbs.js]: Actually, `/_api/data/movies/` returns result for the `/movies/` folder and `/_api/data/movies` returns for `movies.yml` file, both being valid data source in their own context and both resulting in `site.data.movies`

File: src/containers/views/DataFileEdit.js [hunk 5]
```diff
@@ -95,11 +118,16 @@ export class DataFileEdit extends Component {
...
-          <Breadcrumbs splat={filename} type="datafiles" />
+          <Breadcrumbs splat={directory || ""} type="data files" />
         </div>
 
         <div className="content-wrapper">
           <div className="content-body">
+            <InputPath
+              onChange={onDataFileChanged}
+              type="data files"
+              path={path} // TODO: Support using `relative_path` from API instead
```
Comment (mertkahyaoglu) [inline src/containers/views/DataFileEdit.js]: For pages and documents, only `filenames` are passed to `InputPath` component. Relative paths are shown in the `Breadcrumb` component. We should be consistent with them. I'm testing locally and seeing `/_data/data_file.yml` instead for InputPath. Before the filename is already shown to users in the Breadcrumbs component. No need to repeat it. Let's abstract that.

File: src/components/Breadcrumbs.js (hunk 1 above)
Comment (mertkahyaoglu) [inline src/components/Breadcrumbs.js]: But we defined our routes so that individuals are called with their extensions. 

```ruby
get "/*?/?:path.?:ext?"
```

`/_api/data/movies` returning for `movies.yml` file doesn't make sense for me. IMO we need to be as specific as possible while calling the endpoints.

File: src/components/form/InputPath.js (hunk 4 above)
Comment (ashmaroli) [inline src/components/form/InputPath.js]: actually, it was intended to show that `directory` and `/` are optional. The `/` needed only if there was a directory param.
```
(directory/)?filename.yml
(directory/)filename.yml
[directory/]filename.yml
<directory/>filename.yml
```

File: src/components/Breadcrumbs.js (hunk 1 above)
Comment (ashmaroli) [inline src/components/Breadcrumbs.js]: > /_api/data/movies returning for movies.yml file doesn't make sense for me.

It didn't make sense to me too till I read the spec files.

File: src/components/form/InputPath.js (hunk 4 above)
Comment (mertkahyaoglu) [inline src/components/form/InputPath.js]: I didn't get it. I don't think non-technical users will do 😬  HTML tag came to my mind directly.

File: src/containers/views/DataFileEdit.js (hunk 5 above)
Comment (ashmaroli) [inline src/containers/views/DataFileEdit.js]: `/_data/data_file.yml` is shown because that parameter (`path`) is required to be changed to rename the data file.

File: src/components/Breadcrumbs.js (hunk 1 above)
Comment (mertkahyaoglu) [inline src/components/Breadcrumbs.js]: I'm suggesting to fix it since we are proposing sub-directory support now.
@benbalter What do you think?

File: src/components/form/InputPath.js (hunk 4 above)
Comment (ashmaroli) [inline src/components/form/InputPath.js]: Ah! I see. I'll settle for the your suggestion then `<directory> / your-filename.yml`

File: src/components/Breadcrumbs.js (hunk 1 above)
Comment (ashmaroli) [inline src/components/Breadcrumbs.js]: > But we defined our routes so that individuals are called with their extensions.

Its not so according to [`master`](https://github.com/jekyll/jekyll-admin/blob/master/lib/jekyll-admin/server/data.rb#L11)
```ruby
get "/:path.?:ext?" do
```
(I vote for the more sensible resolution too)

Comment (mertkahyaoglu) [inline src/components/Breadcrumbs.js]: Yeah, I just noticed that :face_palm: :trollface:

Comment (mertkahyaoglu) [inline src/components/Breadcrumbs.js]: I think it was a flexible solution at the time, props to @benbalter 🎉  Directories change everything 😄

File: src/containers/views/DataFileEdit.js (hunk 5 above)
Comment (ashmaroli) [inline src/containers/views/DataFileEdit.js]: renaming `/_data/data_file.yml` to `/_data/foo/data_file.yml` will **move** the data file into a subdirectory named `foo`.

Comment (mertkahyaoglu) [inline src/containers/views/DataFileEdit.js]: Yeah, that's expected. What I'm saying is just show `data_file.yml` like posts, for example. You can rename it to `foo/data_file.yml` and it will drop into `/_data/foo/data_file.yml` automagically 😄   Just check out other types, you will understand the difference.

Comment (ashmaroli) [inline src/containers/views/DataFileEdit.js]: I did check other types. Hence added the `//TODO` here for a future PR

Comment (ashmaroli) [inline src/containers/views/DataFileEdit.js]: With the current behavior, you can convert the data file into a static file by dropping it outside the `data_dir` :grinning: 
```
/_data/data_file.yml => /foo/data_file.yml
```

Comment (mertkahyaoglu) [inline src/containers/views/DataFileEdit.js]: You just prepend `_data` to `path` which is the concatenation of the current directory and the relative path that comes from the input in `putDatafile` action. [Here](https://github.com/jekyll/jekyll-admin/blob/master/src/actions/collections.js#L109) is how I managed to do that for collections that won't drop into `pages`.

Comment (ashmaroli) [inline src/containers/views/DataFileEdit.js]: but `_data` will change based on configuration. We need to prepend `DataFile.data_dir` instead. But that is currently not exposed via API

Comment (mertkahyaoglu) [inline src/containers/views/DataFileEdit.js]: That info is already exposed via `path`, `relative_path` and `api_url`. We just need to parse it.

Comment (ashmaroli) [inline src/containers/views/DataFileEdit.js]: On `master`, `path == relative_path` for this resource.
In this PR, they are redefined.
So, do we parse it like below ?

``` javascript
const data_dir = path - relative_path
```

Comment (mertkahyaoglu) [inline src/containers/views/DataFileEdit.js]: Yeah, that should work.

Comment (ashmaroli) [inline src/containers/views/DataFileEdit.js]: what do you think about enabling the tooltip for `DataFileEdit`? If yes, what should the message be?

Comment (mertkahyaoglu) [inline src/containers/views/DataFileEdit.js]: I can't think of one.

**Comment by ashmaroli** [comment]
reinstated `getDirectoryFromPath()` for `DataFiles` and `DataFileNew` as `[directory, ...rest] = params.splat` returned bugs on manual testing.

**Comment by ashmaroli** [comment]
the tests fail as `refs` are not defined for `shallow` wrappers. Trying to `mount` locally throws again at `#l35`: `router.setRouteLeaveHook()`

**Comment by mertkahyaoglu** [comment]
Two things that bothers me.. 😄 

1. screen shot 2017-05-12 at 18 24 31

2. screen shot 2017-05-12 at 18 24 35

**Comment by ashmaroli** [comment]
@mertkahyaoglu Assume I changed it. Now, how would you be **move** `_data/movies/genres/fiction.yml` to `_data/books/favs/fiction.yml` via the `DataFileEdit` container?

**Comment by ashmaroli** [comment]
@benbalter can I remove API support to access `_api/data/foo.yml` via `_api/data/foo`?

**Comment by mertkahyaoglu** [comment]
> how would you move _data/movies/genres/fiction.yml to _data/books/favs/fiction.yml via the DataFileEdit container?

I intentionally disabled this functionality when I was implementing directory support for `Pages` and `Collections`. I don't think it's a good idea to give users too much flexibility/power in the front-end. They can mistakenly move a file to another directory or maybe another content type or even out of Jekyll site and it is really troublesome to move it back.

IMO, these type of operations should be done via OS, not via the GUI. And also I don't want to bring extra complexity to the codebase. This feature requires a bit too much validation to deal with for now. In the future, when Jekyll Admin can be hosted on the cloud, we may implement this or maybe drag-and-drop functionality for files.

Finishing off by quoting from Ben Balter: Sometimes removing/omitting a feature is a feature ☯️

**Comment by ashmaroli** [comment]
Alrite then, I'll make the change.. funny, that field's label says `Path:` when it should rightly be `Filename`..

Regarding the other concerns, directory traversal is already delimited by the API. The `put` action will always be within `JekyllAdmin::Datafile.data_dir`, and other content-types are possible even at current master by simply changing the extension in the `InputPath`. That guard has been provided by the `DataGUI` though..


> Sometimes removing/omitting a feature is a feature

Cool quote! But it doesn't send a clear answer: `yes` or `no`

**Comment by mertkahyaoglu** [comment]
> funny, that field's label says Path: when it should rightly be Filename

Nope, you can go deeper from the current directory by passing `test/othertest/page.md` to `InputPath` but you can't go other directions. Therefore `Path` makes sense, I think..

**Comment by ashmaroli** [comment]
@mertkahyaoglu addressed the two things that bothered you.. :smile:

File: lib/jekyll-admin/server/data.rb [hunk 6]
```diff
@@ -4,16 +4,17 @@ class Server < Sinatra::Base
     EXTENSIONS = %w(yml json).freeze
 
     namespace "/data" do
-      get do
-        json DataFile.all.map(&:to_api)
-      end
-
-      get "/:path.?:ext?" do
+      get "/*?/?:path.?:ext?" do
```
Comment (mertkahyaoglu) [inline lib/jekyll-admin/server/data.rb]: Let's go with `/*?/?:path.:ext` 👍

File: lib/jekyll-admin/server/data.rb [hunk 7]
```diff
@@ -23,7 +24,7 @@ class Server < Sinatra::Base
         json written_file.to_api(:include_content => true)
       end
 
-      delete "/:path.?:ext?" do
+      delete "/*?/?:path.?:ext?" do
```
Comment (mertkahyaoglu) [inline lib/jekyll-admin/server/data.rb]: Same here.

File: lib/jekyll-admin/server/data.rb (hunk 6 above)
Comment (ashmaroli) [inline lib/jekyll-admin/server/data.rb]: so to confirm, we're dropping support for accessing extension-less data files?

Comment (mertkahyaoglu) [inline lib/jekyll-admin/server/data.rb]: Yeap.

Comment (ashmaroli) [inline lib/jekyll-admin/server/data.rb]: k.. so change frontend to send requests to `/data/foo-dir` or leave it at `/data/foo-dir/`?

File: src/containers/views/DataFileNew.js [hunk 8]
```diff
@@ -117,7 +117,7 @@ export class DataFileNew extends Component {
       datafileChanged, fieldChanged, onDataFileChanged, datafile, updated, errors, params
     } = this.props;
     const { path, raw_content } = datafile;
-    const initialPath = params.splat ? (params.splat + "/") : "";
+    const initialPath = params.splat ? (params.splat) : "";
```
Comment (mertkahyaoglu) [inline src/containers/views/DataFileNew.js]: This is easier.
```javascript
const initialPath = params.splat || '';
```

Comment (mertkahyaoglu) [inline src/containers/views/DataFileNew.js]: Actually this is not used at all.

File: src/containers/views/DataFiles.js [hunk 9]
```diff
@@ -30,14 +30,13 @@ export class DataFiles extends Component {
   handleClickDelete(path) {
     const { deleteDataFile, params } = this.props;
     const confirm = window.confirm(getDeleteMessage(path));
-    const directory = params.splat ? (params.splat.replace(/\/$/, "")) : "";
-    const goTo = params.splat ? (params.splat) : "";
+    const directory = params.splat ? (params.splat) : "";
```
Comment (mertkahyaoglu) [inline src/containers/views/DataFiles.js]: Again.
```javascript
const initialPath = params.splat || '';
```

Comment (ashmaroli) [inline src/containers/views/DataFiles.js]: Ah yes.. its a remnant from an earlier commit. One cant call `replace` on a `nil` entity.. 
will change..

File: src/containers/views/DataFiles.js [hunk 10]
```diff
@@ -1,28 +1,43 @@
...
+    if (this.props.params.splat !== nextProps.params.splat) {
+      fetchDataFiles(nextProps.params.splat);
+    }
+  }
+
+  handleClickDelete(path) {
+    const { deleteDataFile, params } = this.props;
+    const confirm = window.confirm(getDeleteMessage(path));
+    const directory = params.splat || "";
+    const dir = directory ? `/${directory}` : ""
```
Comment (mertkahyaoglu) [inline src/containers/views/DataFiles.js]: Semicolon here 😄

**Comment by mertkahyaoglu** [comment]
@ashmaroli What do you think about type converting? Let's say I created a JSON data file first, then decided to switch to YAML using the selectbox. Currently it's not converting. Shall we support it or expect users to decide first then input their data accordingly? 😄

File: src/containers/views/DataFileEdit.js [hunk 11]
```diff
@@ -51,51 +74,126 @@ export class DataFileEdit extends Component {
...
+    obj[key] = value;
+
+    this.setState(obj);
+    onDataFileChanged();
   }
 
   handleClickSave(e) {
     const { datafile, datafileChanged, fieldChanged, putDataFile, params } = this.props;
+    const { path, relative_path } = datafile;
+    const data_dir = path.replace(relative_path, "");
```
Comment (mertkahyaoglu) [inline src/containers/views/DataFileEdit.js]: Try to use single quote for convention please. I also do that 😄

File: src/containers/views/DataFiles.js (hunk 10 above)
Comment (ashmaroli) [inline src/containers/views/DataFiles.js]: aargh.. I hate it that the linter doesn't report this while running `npm run test:watch`

File: src/containers/views/DataFileEdit.js (hunk 11 above)
Comment (ashmaroli) [inline src/containers/views/DataFileEdit.js]: the convention I saw used in all `container/views/*` is that single quotes are used at the top when importing components, and double quotes within the current component class body.
So I think its best to handle all those files in a separate dedicated PR..

**Comment by ashmaroli** [comment]
> What do you think about type converting? Let's say I created a JSON data file first, then decided to switch to YAML using the selectbox. Currently it's not converting. Shall we support it..

I had added support for it via the DataGUI.. but checking now, it seems like minor refactors I made to the `putDatafile` action across various PRs introduced an uncaught bug where the type conversion only takes place when  altering the `basename` and `ext` for the said file. I'll fix it and extend support to the raw editor as well..

File: src/containers/views/DataFileEdit.js (hunk 11 above)
Comment (mertkahyaoglu) [inline src/containers/views/DataFileEdit.js]: Yeah. I also forget them sometimes 😬  Just a reminder for future commitments. Appreciate for the PR though 👍

Comment (ashmaroli) [inline src/containers/views/DataFileEdit.js]: Let me know if that PR will get merged.. I can then correct files here as part of resolving conflicts

Comment (mertkahyaoglu) [inline src/containers/views/DataFileEdit.js]: I was planning to merge it last.

Comment (ashmaroli) [inline src/containers/views/DataFileEdit.js]: okay then..

**Comment by mertkahyaoglu** [comment]
@ashmaroli One last thing. Could you move `Create` button above `Switch` button for DataFileNew?

**Comment by ashmaroli** [comment]
Sorry for the long list of commits :innocent: 
Thank you @mertkahyaoglu for the time and patient review :smiley:

**Comment by mertkahyaoglu** [comment]
@ashmaroli Glad to work with you 😄 

Should've hit `squash and merge` 😬
